File: frontend/src/shared/types/source.ts
export type SourceType = "url" | "text" | "pdf";
export type SourceStatus = "pending" | "processing" | "ready" | "failed";
export type JobStage = "fetching" | "extracting" | "cleaning" | "chunking" | "saving" | "analyzing";
export type JobStatus = "queued" | "running" | "completed" | "failed" | "cancelled";

export type ProcessingJob = {
  id: string;
  source_id: string;
  status: JobStatus;
  stage: JobStage;
  progress: number;
  attempt_count: number;
  error_code: string | null;
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
  updated_at: string;
};

export type Source = {
  id: string;
  workspace_id: string;
  type: SourceType;
  url: string | null;
  title: string;
  status: SourceStatus;
  content_hash: string | null;
  language: string | null;
  created_at: string;
  updated_at: string;
  job: ProcessingJob | null;
};

export type SourceInput = {
  workspaceId: string;
  type: SourceType;
  title: string;
  url?: string;
  text?: string;
  file?: File;
};

File: backend/app/schemas/source.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

SourceType = Literal["url", "pdf", "text"]
SourceStatus = Literal["pending", "processing", "ready", "failed"]
JobStatus = Literal["queued", "running", "completed", "failed", "cancelled"]
JobStage = Literal["fetching", "extracting", "cleaning", "chunking", "saving", "analyzing"]


class ProcessingJobRead(BaseModel):
    id: UUID
    source_id: UUID
    status: JobStatus
    stage: JobStage
    progress: int
    attempt_count: int
    error_code: str | None
    error_message: str | None
    started_at: datetime | None
    finished_at: datetime | None
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SourceRead(BaseModel):
    id: UUID
    workspace_id: UUID
    type: SourceType
    url: str | None
    title: str
    status: SourceStatus
    content_hash: str | None
    language: str | None
    created_at: datetime
    updated_at: datetime
    job: ProcessingJobRead | None = None


class ChunkRead(BaseModel):
    id: UUID
    position: int
    heading_path: str
    content: str
    token_count: int

    model_config = ConfigDict(from_attributes=True)


class DocumentRead(BaseModel):
    id: UUID
    version: int
    clean_content: str
    content_hash: str | None
    word_count: int
    chunks: list[ChunkRead]

    model_config = ConfigDict(from_attributes=True)


class SourceDetail(SourceRead):
    document: DocumentRead | None = None
